docs/design_proposals/sync-auto-support.md

```diff
@@ -0,0 +1,213 @@
+4. run `./gradlew classes`
+5. *skaffold syncs files*
+
+which is far from ideal.
+
+#### Fix it?
+
+What we might want, is an API that can accept configuration that looks like:
+
+##### User Configuration
+
+User's that use `auto` should expect the builder-sync to do the right thing and
+will not be required to do much configuration.
+
+Auto will only work with builders that have implemented the auto spec. We
+expect at least `jib` to do this.
+
+```yaml
+build:
+  artifacts:
+  - image: ...
+    context: jib-project
+    jib: {}
+    sync:
+      auto: {}
+```
+
+##### Sync Api with Buidler
+
+The builder will talk to sync component by providing it with the following data
+1. A list of `generated` configs, each containing
+    1. A `command` to generate files to sync
+    2. A list of inputs to watch as triggers
+    3. A list of syncs (src, dest) to execute after generation
+2. A list of `direct` sync directives (src, dest) to execute without any script
+   execution
+
+So maybe some datastructures like (I don't really know a lot of go, so assume
+this will be written in some consistent way eventually):
+
+```
+type AutoSync struct {
+  generated []Generated
+  direct []Direct
+}
+
+type Generated struct {
+  command []String
+  inputs []File
+  syncables []Syncables
+}
+
+type Direct struct {
+  syncables []Syncables
+}
+
+type Syncables struct {
+  src String
+  dest String
+  strip String
+}
+```
+
+###### Jib - Skaffold Sync interface
+
+How a tool like Jib might surface the necessary information to Skaffold
+
+I would expect to add a task like `_jibSkaffoldSyncMap` that will produce
+json output for the skaffold-jib intergration to consume and forward to the sync
+system. And example output might look like:
+
+```
+BEGIN JIB JSON
+{
+  "generated": {
+    “target/classes”: "app/classes",
+    "target": "dest2"
+  },
+  "direct": {
+    "src/main/extra1": "/",
+    "src/main/extra2": "/"
+    ".m2/some/dep/my-dep-SNAPSHOT.jar": "app/dependencies/my-dep-SNAPSHOT.jar"
+  }
+}
+```
+
+perhaps jib should also tell the system what task/goals are required to build
+the generated files??
+
+```
+{
+  ...
+  "generator-tasks": ["classes", "resources"]
+  "generated": ...
+}
+```
+
+
+### Open Issues/Questions
+
+Please list any open questions here in the following format:
+
+**What happens if the build definition changes?**
+
+I'm not sure if we should just allow the user to go nuts, or if we should
+explicitly disallow this, because changing the build definition might result in
+a post-sync state that is broken.
+
+** What about files that have specific permissions? **
+Jib allows users to customize file permissions, for instance a file on the
+container can be configured to be 0x456 or something. One option instead of
+dealing with this, is to just make all sync'd files 777? Or we allow passthrough
+of permissions from the build system to the sync system.
+
+**Should we allow the user to configure the auto block?**
```

Comment:
This is an interesting question. Just like the user may want to ignore some files handled by "auto", they may want to sync files not synced. But if we don't allow any, "auto" becomes quite restricted, and the user is left with no good option (i.e., the status quo) for the dev loop scenario we are talking about if they wanted to add or remove extra sync rules.

At the same time, it would be difficult for the user to know what files a builder auto-syncs to where unless they are well documented, so manually adding extra "ignored" or "syncables" may not be straightforward.